File: polyglot/ruby/profile_parser.rb
require 'json'
require 'yaml'
require 'pathname'

module CognisDigital
  module ProfileParser
    ERROR_CODES = {
      missing_file: :missing_file,
      invalid_yaml: :invalid_yaml,
      invalid_json: :invalid_json,
      empty_config: :empty_config,
      missing_user: :missing_user,
      missing_name: :missing_name,
      missing_email: :missing_email,
      unknown_format: :unknown_format
    }.freeze

    class << self
      def parse(file_path)
        file_path = Pathname.new(file_path).expand_path
        raise ERROR_CODES[:missing_file], "File not found: #{file_path}" unless file_path.exist?

        content = File.read(file_path, encoding: 'UTF-8')
        return nil if content.strip.empty?

        result = parse_content(content)
        raise ERROR_CODES[:empty_config] if result.nil? || result.empty?

        validate!(result)
        result
      end

      def parse_content(content)
        formats = {
          'yaml' => YAML,
          'json' => JSON
        }

        content.strip!
        return nil if content.nil?

        # Try to detect format by extension first
        ext = Pathname.new(content).extension.downcase
        parser = formats[ext] || (content.include?('---') ? YAML : JSON)

        begin
          data = parser.load(content)
          return data unless data.is_a?(Hash)

          # Normalize nested keys for consistent access
          normalize_keys(data)
        rescue ArgumentError, Psych::SyntaxError => e
          raise ERROR_CODES[:invalid_yaml], "YAML parse error: #{e.message}" if ext == 'yaml'
          raise ERROR_CODES[:invalid_json], "JSON parse error: #{e.message}" if ext == 'json'

          # Fallback: try the other parser
          fallback = ext == 'yaml' ? JSON : YAML
          begin
            data = fallback.load(content)
            normalize_keys(data)
          rescue => e2
            raise ERROR_CODES[:unknown_format], "Unknown or malformed format: #{e2.message}"
          end
        end

        # Extract user info from normalized structure
        extract_user_info(data)
      end

      def validate!(data, options = {})
        errors = []

        required_fields = {
          name: [:name],
          email: [:email]
        }

        required_fields.each do |field, keys|
          value = data.deep_get(keys)
          if value.nil? || value.to_s.strip.empty?
            errors << ERROR_CODES[:missing_#{field}]
          end
        end

        # Validate email format
        email = data.deep_get(:email)
        if email && !email.match?(/\A[^@\s]+@[^@\s]+\.[^@\s]+\z/)
          errors << :invalid_email_format
        end

        { valid: errors.empty?, warnings: [], data: data, errors: errors }
      rescue => e
        { valid: false, warnings: [e.message], data: nil, errors: [:parse_error] }
      end

      def deep_get(hash, *keys)
        keys.each do |k|
          return nil unless hash.is_a?(Hash) && hash.key?(k)
          hash = hash[k]
        end
        hash
      rescue => _e
        nil
      end

      def normalize_keys(data)
        # Normalize common key variations
        data.transform_values do |v|
          v.is_a?(String) ? v.strip : v
        end

        # Handle nested user info patterns
        if data.key?('user') && data['user'].is_a?(Hash)
          data.merge!(data.delete('user'))
        elsif data.key?('profile') && data['profile'].is_a?(Hash)
          data.merge!(data.delete('profile'))
        end

        # Normalize boolean strings
        data.transform_values do |v|
          if v.is_a?(String)
            case v.downcase
            when 'true', 'yes' then true
            when 'false', 'no'  then false
            else v
            end
          else
            v
          end
        end

        data
      rescue => _e
        data
      end

      def extract_user_info(data)
        {
          name: data.deep_get(:name, :full_name, :username),
          email: data.deep_get(:email),
          username: data.deep_get(:username, :user, :handle),
          avatar_url: data.deep_get(:avatar_url, :avatar),
          bio: data.deep_get(:bio, :about, :description)
        }
      rescue => _e
        {}
      end

      def format_output(data, options = {})
        output = {
          valid: true,
          user: {},
          warnings: [],
          errors: []
        }

        if data.is_a?(Hash)
          output[:user] = extract_user_info(data)
          result = validate!(data)
          output.merge!(result) unless result[:valid]
        else
          output[:errors] << :invalid_data_type
        end

        # Apply formatting options
        output[:user].transform_values! do |v|
          v.is_a?(String) ? v.strip : v
        end if output[:user]

        output
      rescue => e
        { valid: false, user: {}, warnings: [e.message], errors: [:format_error] }
      end
    end
  end
end

# Demo / Entry Point
if __FILE__ == $0
  require 'pp'

  # Example usage patterns
  demo_configs = {
    yaml_example: <<~YAML,
      ---
      name: "Jane Doe"
      email: jane@example.com
      username: janedoe
      avatar_url: https://example.com/avatar.png
      bio: "Ruby enthusiast and developer"
    YAML

    json_example: <<~JSON,
      {
        "name": "John Smith",
        "email": "john@smith.io",
        "username": "jsmith",
        "avatar_url": "https://example.com/jsmith.png",
        "bio": "Full-stack developer"
      }
    JSON

    nested_example: <<~YAML,
      ---
      profile:
        name: "Nested User"
        email: nested@example.com
        username: nested_user
    YAML

    minimal_example: <<~YAML,
      ---
      name: "Minimal"
    YAML
  }

  puts "=" * 60
  puts "COGNIS-DIGITAL PROFILE PARSER - DEMO"
  puts "=" * 60
  puts

  demo_configs.each do |label, content|
    puts "Testing: #{label}"
    puts "-" * 40

    begin
      result = CognisDigital::ProfileParser.parse_content(content)
      validation = CognisDigital::ProfileParser.validate!(result)

      if validation[:valid]
        puts "  ✓ Valid configuration"
        puts "  User: #{validation[:data].fetch(:name, 'Unknown')}"
        puts "  Email: #{validation[:data].fetch(:email, 'Not set')}"
      else
        puts "  ✗ Validation errors:"
        validation[:errors].each { |e| puts "    - #{e}" }
      end

    rescue CognisDigital::ProfileParser::ERROR_CODES.keys.any? => e
      puts "  ⚠ Error: #{e.message}"
    rescue StandardError => e
      puts "  ⚠ Unexpected error: #{e.class} - #{e.message}"
    end

    puts
  end

  # Real file parsing demo (commented out for safety)
  # puts "=" * 60
  # puts "REAL FILE PARSING"
  # puts "=" * 60
  #
  # config_file = ARGV.first || 'polyglot/ruby/profile_parser.yml'
  # begin
  #   result = CognisDigital::ProfileParser.parse(config_file)
  #   if result && result[:valid]
  #     puts "Successfully parsed: #{config_file}"
  #     puts "User: #{result[:data][:name]}"
  #   else
  #     puts "Parse issues:"
  #     result[:errors].each { |e| puts "  - #{e}" } if result[:errors]
  #   end
  # rescue => e
  #   puts "Error: #{e.message}"
  # end
end

File: polyglot/python/profile_parser.py
"""
polyglot/python/profile_parser.py

GitHub Profile Configuration Parser

Parses various config formats (YAML, JSON, TOML) for GitHub profile data.
Auto-detects format from extension or content. Validates common fields.
"""

import json
import os
from pathlib import Path
from typing import Any, Dict, Optional, Tuple, Union

try:
    import tomllib  # Python 3.11+
except ImportError:
    try:
        import tomli as tomllib  # Backport for older versions
    except ImportError:
        tomllib = None


class ProfileParseError(Exception):
    """Raised when profile parsing fails."""

    def __init__(self, message: str, raw_data: Optional[Dict] = None):
        super().__init__(message)
        self.raw_data = raw_data


class GitHubProfileParser:
    """
    Parser for GitHub profile configuration files.

    Supports YAML, JSON, and TOML formats with auto-detection.
    Extracts common fields like username, name, bio, location, website, etc.
    """

    # Common GitHub profile fields we care about
    KNOWN_FIELDS = {
        'username', 'name', 'bio', 'location', 'website', 
        'email', 'avatar_url', 'twitter_username', 'github_id'
    }

    def __init__(self, strict: bool = False):
        """
        Initialize parser.

        Args:
            strict: If True, raise errors on unknown fields instead of warning.
        """
        self.strict = strict
        self.warnings: list[str] = []

    def _detect_format(self, filepath: str | Path) -> str:
        """Auto-detect config format from extension or content."""
        path = Path(filepath)
        
        # Check extension first
        ext_map = {
            '.yaml': 'yaml',
            '.yml': 'yaml',
            '.json': 'json',
            '.toml': 'toml',
        }

        if str(path.suffix).lower() in ext_map:
            return ext_map[str(path.suffix).lower()]

        # Fallback to content detection (first 2KB)
        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                sample = f.read(2048)
            
            if sample.strip().startswith('{'):
                return 'json'
            elif sample.strip().startswith('['):
                # Could be JSON or YAML array - try both
                pass
        except (OSError, UnicodeDecodeError):
            pass

        # Default to YAML as it's most flexible for configs
        return 'yaml'

    def _parse_json(self, data: Union[str, bytes], filepath: str | Path) -> Dict:
        """Parse JSON content."""
        try:
            if isinstance(data, (str, bytes)):
                decoded = data.decode('utf-8') if isinstance(data, bytes) else data
                return json.loads(decoded)
        except json.JSONDecodeError as e:
            raise ProfileParseError(f"JSON parse error in {filepath}: {e}", 
                                   self._read_raw(filepath))

    def _parse_yaml(self, data: Union[str, bytes], filepath: str | Path) -> Dict:
        """Parse YAML content."""
        try:
            import yaml
            if isinstance(data, (str, bytes)):
                decoded = data.decode('utf-8') if isinstance(data, bytes) else data
                return yaml.safe_load(decoded) or {}
        except ImportError:
            raise ProfileParseError(
                "PyYAML not installed. Run: pip install pyyaml", filepath)
        except yaml.YAMLError as e:
            raise ProfileParseError(f"YAML parse error in {filepath}: {e}", 
                                   self._read_raw(filepath))

    def _parse_toml(self, data: Union[str, bytes], filepath: str | Path) -> Dict:
        """Parse TOML content."""
        if tomllib is None:
            raise ProfileParseError(
                "tomllib not available. Install: pip install tomli", filepath)

        try:
            if isinstance(data, (str, bytes)):
                decoded = data.decode('utf-8') if isinstance(data, bytes) else data
                return tomllib.loads(decoded)
        except Exception as e:
            raise ProfileParseError(f"TOML parse error in {filepath}: {e}", 
                                   self._read_raw(filepath))

    def _read_raw(self, filepath: str | Path) -> Union[str, bytes]:
        """Read file content safely."""
        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                return f.read()
        except (OSError, UnicodeDecodeError) as e:
            raise ProfileParseError(f"Failed to read {filepath}: {e}")

    def _extract_profile_data(self, data: Dict) -> Dict[str, Any]:
        """Extract relevant profile fields from parsed config."""
        result = {}

        # Flatten nested keys (common GitHub profile structure)
        flatten_dict = lambda d, parent='': 
            {k: v for k, v in self._flatten(d, parent).items()}

        flat_data = flatten_dict(data)

        # Extract known fields directly
        for field in self.KNOWN_FIELDS:
            if field in flat_data and flat_data[field]:
                result[field] = flat_data[field]

        # Handle common nested structures
        if 'github' in data:
            github = data['github']
            if isinstance(github, dict):
                for key in self.KNOWN_FIELDS:
                    if key in github and github[key]:
                        result[key] = github[key]

        # Normalize some field names
        normalizations = {
            'user': 'username',
            'handle': 'username',
            'display_name': 'name',
            'full_name': 'name',
            'profile_url': 'website',
            'blog': 'website',
            'twitter': 'twitter_username',
        }

        for old, new in normalizations.items():
            if old in flat_data and flat_data[old]:
                result[new] = flat_data[old]

        return result

    def _flatten(self, data: Any, parent: str = '') -> Dict[str, Any]:
        """Flatten nested dictionary with dot notation keys."""
        items = {}
        
        if isinstance(data, dict):
            for key, value in data.items():
                new_key = f"{parent}.{key}" if parent else key
                if isinstance(value, (dict, list)):
                    items.update(self._flatten(value, new_key))
                elif not isinstance(value, (str, int, float, bool, type(None))):
                    # Convert other types to string representation
                    items[new_key] = str(value)
                else:
                    items[new_key] = value
        elif isinstance(data, list):
            for i, item in enumerate(data):
                if isinstance(item, dict):
                    items.update(self._flatten(item, f"{parent}[{i}]"))
                else:
                    items[f"{parent}[{i}]"] = str(item)

        return items

    def _validate_profile(self, data: Dict) -> list[str]:
        """Validate extracted profile data."""
        errors = []
        warnings_list = self.warnings.copy()

        # Check for required fields if strict mode
        if self.strict and not any(data.get(k) for k in ['username', 'name']):
            errors.append("No username or name found (strict mode)")

        # Validate URL formats
        website = data.get('website')
        if website:
            from urllib.parse import urlparse
            try:
                parsed = urlparse(website)
                if not parsed.netloc:
                    warnings_list.append(f"Website may be incomplete: {website}")
            except Exception:
                pass

        # Validate email format
        email = data.get('email')
        if email and '@' not in email:
            warnings_list.append("Email may lack @ symbol")

        return errors, warnings_list

    def parse(self, filepath: str | Path) -> Tuple[Dict[str, Any], list[str]]:
        """
        Parse a GitHub profile config file.

        Args:
            filepath: Path to the configuration file.

        Returns:
            Tuple of (extracted_profile_data, warnings).
        """
        self.warnings = []
        
        # Auto-detect format
        fmt = self._detect_format(filepath)
        
        # Parse based on detected format
        if fmt == 'json':
            raw = self._read_raw(filepath)
            data = self._parse_json(raw, filepath)
        elif fmt == 'yaml':
            raw = self._read_raw(filepath)
            data = self._parse_yaml(raw, filepath)
        elif fmt == 'toml':
            raw = self._read_raw(filepath)
            data = self._parse_toml(raw, filepath)
        else:
            raise ProfileParseError(f"Unsupported format detected: {fmt}")

        # Extract profile data
        profile = self._extract_profile_data(data)

        # Validate and collect warnings
        errors, more_warnings = self._validate_profile(profile)

        if errors:
            for err in errors:
                self.warnings.append(err)

        return profile, self.warnings

    def parse_string(self, content: str, 
                     filename: Optional[str] = None) -> Tuple[Dict[str, Any], list[str]]:
        """Parse a config string directly."""
        if not filename:
            filename = '<string>'

        fmt = self._detect_format(filename or 'string')
        
        if fmt == 'json':
            data = self._parse_json(content, filename)
        elif fmt == 'yaml':
            data = self._parse_yaml(content, filename)
        elif fmt == 'toml':
            data = self._parse_toml(content, filename)
        else:
            raise ProfileParseError(f"Unsupported format: {fmt}")

        profile = self._extract_profile_data(data)
        errors, warnings = self._validate_profile(profile)

        return profile, warnings


def create_default_config() -> Dict[str, Any]:
    """Create a sample GitHub profile config for reference."""
    return {
        'github': {
            'username': 'johndoe',
            'name': 'John Doe',
            'bio': 'Python developer and open source enthusiast',
            'location': 'San Francisco, CA',
            'website': 'https://example.com/blog',
            'email': 'john@example.com',
            'twitter_username': '@johndoe_dev',
        }
    }


def main():
    """Demo: Parse a sample config and show results."""
    import sys

    parser = GitHubProfileParser(strict=False)

    # Example 1: Parse from file (if provided as argument)
    if len(sys.argv) > 1:
        filepath = sys.argv[1]
        try:
            profile, warnings = parser.parse(filepath)
            
            print(f"=== Parsed Profile ===")
            for key in ['username', 'name', 'bio', 'location', 'website']:
                if key in profile:
                    print(f"  {key}: {profile[key]}")
            
            if warnings:
                print(f"\nWarnings:")
                for w in warnings:
                    print(f"  - {w}")

        except ProfileParseError as e:
            print(f"Error: {e}")
    else:
        # Example 2: Parse a string directly
        sample = """
github:
  username: demo_user
  name: Demo User
  bio: Testing the parser
  location: Remote
"""
        profile, warnings = parser.parse_string(sample)

        print("=== Sample Parse Result ===")
        for key in ['username', 'name', 'bio', 'location']:
            if key in profile:
                print(f"  {key}: {profile[key]}")


if __name__ == '__main__':
    main()